src/utils/usePromise.js

```diff
@@ -0,0 +1,110 @@
+/* eslint react/prop-types: 0 */
+// @flow
+import * as React from 'react';
+
+function resolvePromise(promise) {
+  if (typeof promise === 'function') {
+    return promise();
+  }
+
+  return promise;
+}
+
+const states = {
+  pending: 'pending',
+  rejected: 'rejected',
+  resolved: 'resolved',
+  initial: 'initial',
+};
+
+const INITIAL_PENDING = {
+  error: undefined,
+  result: undefined,
+  state: states.pending,
+  lastValue: null,
+};
+
+function reducer(state, action) {
+  switch (action.type) {
+    case states.resolved:
+      return {
+        error: undefined,
+        result: action.payload,
+        state: states.resolved,
+        lastValue: { inner: action.payload },
+      };
+    case states.rejected:
+      return {
+        error: action.payload,
+        result: undefined,
+        state: states.rejected,
+        lastValue: state.lastValue,
+      };
+    case states.pending:
+      return {
+        error: undefined,
+        result: undefined,
+        state: states.pending,
+        lastValue: state.lastValue,
+      };
+
+    case states.initial:
+      return {
+        error: null,
+        result: null,
+        state: states.initial,
+        lastValue: null,
+      };
+    default:
+      return state;
+  }
+}
+
+function usePromise(_promise, deps) {
+  let promise = _promise;
+  const [{
+    result,
+    error,
+    state,
+    lastValue,
+  }, dispatch] = React.useReducer(reducer, INITIAL_PENDING);
+
+  const isFirstRun = React.useRef(true);
+
+  React.useEffect(() => {
+    promise = resolvePromise(promise);
+    let canceled = false;
+
+    if (!isFirstRun.current) {
+      dispatch({ type: states.pending });
+    }
+
+    if (!promise) {
+      dispatch({ type: states.initial });
+      return;
+    }
```

Comment:
I think this could be simplified and more correct as:

```js
if (!promise) {
  dispatch({ type: states.initial });
} else if (state !== states.pending) {
  dispatch({ type: states.pending });
}
```